src/storage/mod.rs

```diff
@@ -864,21 +927,64 @@ impl Storage {
         Ok(())
     }
 
+    fn raw_scan(
+        snapshot: Box<Snapshot>,
+        start_key: &Key,
+        limit: usize,
+        stats: &mut Statistics,
+    ) -> engine::Result<Vec<Result<KvPair>>> {
+        let mut cursor = snapshot.iter(IterOption::default(), ScanMode::Forward)?;
+        if !cursor.seek(start_key, &mut stats.data)? {
+            return Ok(vec![]);
+        }
+        let mut pairs = vec![];
+        while cursor.valid() && pairs.len() < limit {
+            pairs.push(Ok((cursor.key().to_owned(), cursor.value().to_owned())));
+            cursor.next(&mut stats.data);
+        }
+        Ok(pairs)
+    }
+
     pub fn async_raw_scan(
         &self,
         ctx: Context,
         key: Vec<u8>,
         limit: usize,
-        callback: Callback<Vec<Result<KvPair>>>,
-    ) -> Result<()> {
-        let cmd = Command::RawScan {
-            ctx: ctx,
-            start_key: Key::from_encoded(key),
-            limit: limit,
-        };
-        self.schedule(cmd, StorageCb::KvPairs(callback))?;
-        RAWKV_COMMAND_COUNTER_VEC.with_label_values(&["scan"]).inc();
-        Ok(())
+    ) -> Box<Future<Item = Vec<Result<KvPair>>, Error = Error> + Send> {
+        static CMD: &'static str = "raw_scan";
+        let engine = self.get_engine();
+        let priority = readpool::Priority::from(ctx.get_priority());
+
+        let res = self.read_pool.future_execute(priority, move |ctxd| {
+            let _timer = {
+                let ctxd = ctxd.clone();
+                let mut thread_ctx = ctxd.current_thread_context_mut();
+                thread_ctx.collect_command_count_and_duration(CMD, priority, true)
+            };
+
+            Self::async_snapshot(engine, &ctx).and_then(move |snapshot: Box<Snapshot>| {
+                let mut thread_ctx = ctxd.current_thread_context_mut();
+                let _t_process = thread_ctx.collect_processing_read_duration(CMD);
+
+                let mut statistics = Statistics::default();
+                let result =
+                    Storage::raw_scan(snapshot, &Key::from_encoded(key), limit, &mut statistics)
+                        .map_err(Error::from)
+                        .map(|r| {
+                            // TODO: Should we collect statistics even when failed?
+                            thread_ctx.collect_key_reads(CMD, r.len() as u64);
```

Comment:
Also need collect read bytes.

Reply:
I opened issue #2796 for this. We can deal with this later.